src/middle/Fun_kind.ml

```diff
@@ -1,11 +1,29 @@
 open Core_kernel
 
+type 'a suffix = FnPure | FnRng | FnLpdf of 'a | FnTarget
```

Comment:
idk if `FnPure` is the right word for this, every function for reverse mode has the side effect of adding a callback to the global callback stack for calculating the adjoints when calling `grad()`. Mby just `FnDefault`?

https://gcc.gnu.org/onlinedocs/gcc-4.7.2/gcc/Function-Attributes.html

> Many functions have no effects except the return value and their return value depends only on the parameters and/or global variables. Such a function can be subject to common subexpression elimination and loop optimization just as an arithmetic operator would be. These functions should be declared with the attribute pure. For example,

>          int square (int) __attribute__ ((pure));
     
> says that the hypothetical function square is safe to call fewer times than the program says.

> Some of common examples of pure functions are strlen or memcmp. Interesting non-pure functions are functions with infinite loops or those depending on volatile memory or other system resource, that may change between two consecutive calls (such as feof in a multithreading environment). 

I think we reverse mode does not follow the line here

> except the return value and their return value depends only on the parameters and/or global variables.

Since the inputs will also depend on global variables (the global call stack we put the gradient calculations on)

Reply:
They're pure functions in that the autodiff callback stack is not a user-visible side effect. In particular, common subexpression elimination is a legal optimization.
That being said, I'm not a huge fan of either `FnPure` or `FnDefault`. I considered a plain `Fn` but it feels _too_ short.

Reply:
Maybe `FnBasic` or `FnPlain`?

Reply:
`FnPlain`? tbh `Fn` is a bit short but fine by me.

Reply:
^lol I think we commented within a few minutes of each other

Reply:
I'm fairly sure you both "came up with the same idea" only because I literally said "a plain `Fn`" but sure `FnPlain` is as good as anything else I can think of.